### PR TITLE
Fix hdr-dng white balance

### LINKED ISSUE #8043: modern chromatic adaptation clashes with HDR DNG
The defaults settings for colorin.c, upon which the ones of temperature.c and channelmixerrgb.c depend for the white balance, always use the embedded color profile for DNG files. This was changed last december to properly support DNG files. Previously, if available, colorin.c would default to the standard matrix if available first.

When darktable creates an HDR by merging pictures, it doesn't embed the input profile in the resulting DNG. When opening the DNG, the white balance is off, as well as the input profile.

To do :
force the DNG merging to copy and embed the white balance coeffs and the input profile, from the source images, then deal with DNG in an uniform way.

### PR DESCRIPTION
**Pull request by Cobert0**
To get correct White balance for hdr dng, camera matrix is added in hdr dng tiff header : ColorMatrix1 = adobe coeff, so neutral balance is fixed, camera white balance is also added in hdr dng tiff header as AsShotNeutral for rawspeed dng decoder WB, so color calibration can work. Input profile embeded matrix is equal to standard matrix.
Fix #8043

### CODE REVIEW

File: src/common/imageio_dng.h [hunk 1]
```diff
@@ -152,25 +160,41 @@ static inline void dt_imageio_dng_write_tiff_header(
   white.f = whitelevel;
   b = dt_imageio_dng_make_tag(50717, LONG, 1, white.u, b, &cnt); // WhiteLevel in float, actually.
   b = dt_imageio_dng_make_tag(50721, SRATIONAL, 9, 480, b, &cnt); // ColorMatrix1 (XYZ->native cam)
-
-  // b = dt_imageio_dng_make_tag(50728, RATIONAL, 3, 512, b, &cnt); // AsShotNeutral
   // b = dt_imageio_dng_make_tag(50729, RATIONAL, 2, 512, b, &cnt); // AsShotWhiteXY
   b = dt_imageio_dng_make_tag(50778, SHORT, 1, 21 << 16, b, &cnt); // CalibrationIlluminant1
+  b = dt_imageio_dng_make_tag(50728, RATIONAL, 3, 556, b, &cnt); // AsShotNeutral
```
Comment (parafin) [inline src/common/imageio_dng.h:165]: Do not re-arrange tags, they have to go in ascending order (50728 before 50778).

**Review by parafin — CHANGES_REQUESTED** [review]
One comment + there seems to be some indentation problems in the code.

File: src/common/imageio_dng.h (hunk 1 above)
Comment (Cobert0) [reply src/common/imageio_dng.h:165]: Ok , I modify.

**Comment by Nilvus** [comment]
@Cobert0: could you please add "Fix #8043" in your first comment to link the related issue to that PR? That's the only way to automatically link the issue (actually it is not) and not when it is in PR title or commit title like actually. You could check if an issue is linked or not in menu in the right panel of this page. For title, it's better to just precise what the PR is about (and not what issue fixed). Hope I'm clear!

**Comment by TurboGit** [comment]
@Cobert0 : can you please add a meaningful first line to commit log.

`Fix #8043.`

Covey zero information when using git log.

The format should be:

```
Fix bla bla bla ...

Long description about this fix/feature if needed on
multiple line.

Fix #8043.
```

TIA.

**Comment by Cobert0** [comment]
> @Cobert0 : can you please add a meaningful first line to commit log.
> 
> `Fix #8043.`
> 
> Covey zero information when using git log.
> 
> The format should be:
> 
> ```
> Fix bla bla bla ...
> 
> Long description about this fix/feature if needed on
> multiple line.
> 
> Fix #8043.
> ```
> 
> TIA.

Not sure to well understand, is it OK for first comment?

**Comment by TurboGit** [comment]
> Not sure to well understand, is it OK for first comment?

Sorry not sure to understand... "first comment" ?

**Comment by Cobert0** [comment]
> > Not sure to well understand, is it OK for first comment?
> 
> Sorry not sure to understand... "first comment" ?

En français, que dois-je faire?

**Comment by TurboGit** [comment]
Changer le texte du commit. Si tu ne sais pas faire, donne moi un texte descriptif et je ferais le commit.

**Review by TurboGit — CHANGES_REQUESTED** [review]
Style fix needed.

File: src/common/imageio_dng.h [hunk 2]
```diff
@@ -152,25 +160,42 @@ static inline void dt_imageio_dng_write_tiff_header(
...
+  for(int k = 0; k < 9; k++)
+   {
+     dt_imageio_dng_write_buf(buf, 480+k*8, m[k]);
+     dt_imageio_dng_write_buf(buf, 484+k*8, den);
+   }
+
+  for(int k = 0; k < 3; k++)
+   {
+ 	// TAG AsShotNeutral: for rawspeed Dngdecoder camera white balance
+ 	coeff[k] =1/( (wb_coeffs[k]/ wb_coeffs[1])) ;
```
[the remark this answers is not in the document]
Comment (Cobert0) [reply src/common/imageio_dng.h:193]: OK, I update.

**Comment by Cobert0** [comment]
> Changer le texte du commit. Si tu ne sais pas faire, donne moi un texte descriptif et je ferais le commit.

J'ai essayé, je ne sais pas si ok, sinon le texte serait:

fix hdr-dng white balance.
Add camera WB in hdr-dng tiff header as  AsShotNeutral tag.
Add camera matrix from adobe_coeff.c in ColorMatrix1

**Comment by TurboGit** [comment]
@Cobert0 : Note that it is not needed to add the Co-Authored entry in the commit log for yourself. The Co-Authored is meant to add some credit for part of the work to someone else having helped or given a part of the code, etc.

**Comment by aurelienpierre** [comment]
This is better but does not cover all cases, like when input RAWs are already DNG or a custom input profile is used. 

Instead of fetching internal `adobe_coeffs.c`, which is an hard assumption, it would be better to go fetch whatever profile is defined in `colorin.c` module, by using the general input profile. This can be done by using `int dt_colorspaces_get_matrix_from_input_profile()` from `/common/colorspaces.h`

**Comment by Cobert0** [comment]
Ok , I look if I can do it.

Le ven. 19 févr. 2021 à 19:28, Aurélien PIERRE <notifications@github.com> a
écrit :

**Comment by johnny-bit** [comment]
@Cobert0 - Would you have time to look into mentioned by Aurelien? Would be great to respect input matrix instead of defaulting to adobe_coeff ones...

**Comment by Cobert0** [comment]
Hello,
Sorry for my late reply,
My understanding of the software architecture is probably insufficient, I
did not find a simple way to recover the matrix defined in colorin.c from
control_jobs.c.(HDR construction)
On the other hand I don't see how to easily retrieve the original raw
matrix from the D50 matrix of colorin.c without the XYZ white of the
original.
I'm busy at the moment, I'll try to look again.
Sincerely



Le dim. 27 juin 2021 à 21:18, Hubert Kowalski ***@***.***> a
écrit :

**Pull request merged**